File: TI1/TI1.4 RESTFUL_DOTNET_SINBDD_GR09/02.CLIESC/CONUNI_RESTFUL_DOTNET_CLIESC_G09/CONUNI_RESTFUL_DOTNET_CLIESC_G09/vista/PanelLongitud.cs
using System;
using System.Drawing;
using System.Threading.Tasks;
using System.Windows.Forms;
using CONUNI_RESTFUL_DOTNET_CLIESC_G09.servicio;

namespace CONUNI_RESTFUL_DOTNET_CLIESC_G09.vista
{
    /// <summary>
    /// Panel para conversiones de longitud - Diseño compacto y funcional
    /// </summary>
    public class PanelLongitud : Panel
    {
        // Colores oficiales de Monsters Inc.
        private static readonly Color AZUL_PRINCIPAL = Color.FromArgb(33, 169, 218);      // #21a9da
        private static readonly Color ROJO_MONSTER = Color.FromArgb(246, 105, 113);        // #f66971
        private static readonly Color AMARILLO_MONSTER = Color.FromArgb(245, 216, 128);   // #f5d880
        private static readonly Color PURPURA_MONSTER = Color.FromArgb(140, 107, 205);     // #8c6bcd
        private static readonly Color AZUL_CLARO = Color.FromArgb(159, 220, 250);         // #9fdcfa
        private static readonly Color BLANCO = Color.White;
        private static readonly Color GRIS_OSCURO = Color.FromArgb(60, 60, 60);
        private static readonly Color GRIS_CLARO = Color.FromArgb(240, 240, 240);

        // Componentes
        private Label lblTitulo;
        private Label lblValor;
        private TextBox txtValor;
        private Label lblConversion;
        private ComboBox cmbConversion;
        private Button btnConvertir;
        private Label lblResultado;

        // Cliente REST
        private ConversionApiClient clienteREST;

        public PanelLongitud(ConversionApiClient cliente)
        {
            clienteREST = cliente;
            InicializarComponentes();
            ConfigurarLayout();
            ConfigurarEventos();
        }

        private void InicializarComponentes()
        {
            // Título
            lblTitulo = new Label();
            lblTitulo.Text = "CONVERSIONES DE LONGITUD";
            lblTitulo.Font = new Font("Arial", 16F, FontStyle.Bold);
            lblTitulo.ForeColor = GRIS_OSCURO;
            lblTitulo.TextAlign = ContentAlignment.MiddleCenter;

            // Valor
            lblValor = new Label();
            lblValor.Text = "Valor a convertir:";
            lblValor.Font = new Font("Arial", 12F, FontStyle.Bold);
            lblValor.ForeColor = GRIS_OSCURO;

            txtValor = new TextBox();
            txtValor.Font = new Font("Arial", 14F);
            txtValor.TextAlign = HorizontalAlignment.Center;
            txtValor.Size = new Size(200, 30);

            // Conversión
            lblConversion = new Label();
            lblConversion.Text = "Tipo de conversión:";
            lblConversion.Font = new Font("Arial", 12F, FontStyle.Bold);
            lblConversion.ForeColor = GRIS_OSCURO;

            cmbConversion = new ComboBox();
            cmbConversion.Items.AddRange(new string[] {
                "Metros → Pies",
                "Pies → Metros",
                "Metros → Pulgadas",
                "Pulgadas → Metros",
                "Kilómetros → Millas",
                "Millas → Kilómetros"
            });
            cmbConversion.Font = new Font("Arial", 12F);
            cmbConversion.DropDownStyle = ComboBoxStyle.DropDownList;
            cmbConversion.Size = new Size(300, 30);

            // Botón convertir
            btnConvertir = new Button();
            btnConvertir.Text = "CONVERTIR";
            btnConvertir.BackColor = AZUL_PRINCIPAL;
            btnConvertir.ForeColor = BLANCO;
            btnConvertir.Font = new Font("Arial", 16F, FontStyle.Bold);
            btnConvertir.Size = new Size(200, 45);
            btnConvertir.FlatStyle = FlatStyle.Flat;
            btnConvertir.Cursor = Cursors.Hand;

            // Resultado
            lblResultado = new Label();
            lblResultado.Text = "Resultado aparecerá aquí";
            lblResultado.Font = new Font("Arial", 14F);
            lblResultado.ForeColor = GRIS_OSCURO;
            lblResultado.TextAlign = ContentAlignment.MiddleCenter;
            lblResultado.BackColor = GRIS_CLARO;
            lblResultado.AutoSize = false;
            lblResultado.Size = new Size(500, 100);
            lblResultado.BorderStyle = BorderStyle.FixedSingle;
        }

        private Panel panelContenido;
        private Panel panelPrincipal;

        private void ConfigurarLayout()
        {
            this.BackColor = BLANCO;
            this.Dock = DockStyle.Fill;
            this.Padding = new Padding(0);

            // Panel principal para centrar todo
            panelPrincipal = new Panel();
            panelPrincipal.Dock = DockStyle.Fill;
            panelPrincipal.BackColor = BLANCO;

            // Contenedor centrado para el contenido
            panelContenido = new Panel();
            panelContenido.Size = new Size(600, 500);
            panelContenido.BackColor = BLANCO;
            panelContenido.Anchor = AnchorStyles.None;

            // Título centrado
            lblTitulo.Location = new Point(0, 20);
            lblTitulo.Size = new Size(600, 50);
            lblTitulo.TextAlign = ContentAlignment.MiddleCenter;
            panelContenido.Controls.Add(lblTitulo);

            // Valor a convertir
            lblValor.Location = new Point(50, 100);
            lblValor.Size = new Size(200, 30);
            lblValor.TextAlign = ContentAlignment.MiddleLeft;
            panelContenido.Controls.Add(lblValor);

            txtValor.Location = new Point(250, 100);
            txtValor.Size = new Size(300, 35);
            txtValor.Font = new Font("Arial", 12F);
            panelContenido.Controls.Add(txtValor);

            // Tipo de conversión
            lblConversion.Location = new Point(50, 160);
            lblConversion.Size = new Size(200, 30);
            lblConversion.TextAlign = ContentAlignment.MiddleLeft;
            panelContenido.Controls.Add(lblConversion);

            cmbConversion.Location = new Point(250, 160);
            cmbConversion.Size = new Size(300, 35);
            cmbConversion.SelectedIndex = 0;
            panelContenido.Controls.Add(cmbConversion);

            // Botón convertir (centrado)
            btnConvertir.Location = new Point(200, 230);
            btnConvertir.Size = new Size(200, 45);
            btnConvertir.BackColor = AZUL_CLARO;
            btnConvertir.ForeColor = BLANCO;
            btnConvertir.Font = new Font("Arial", 14F, FontStyle.Bold);
            panelContenido.Controls.Add(btnConvertir);

            // Área de resultado
            lblResultado.Location = new Point(50, 310);
            lblResultado.Size = new Size(500, 100);
            lblResultado.BackColor = GRIS_CLARO;
            lblResultado.ForeColor = GRIS_OSCURO;
            lblResultado.BorderStyle = BorderStyle.FixedSingle;
            lblResultado.Text = "Resultado aparecerá aquí";
            panelContenido.Controls.Add(lblResultado);

            // Ajustar posición cuando cambie el tamaño
            panelPrincipal.Resize += PanelPrincipal_Resize;
            this.Resize += PanelPrincipal_Resize;

            panelPrincipal.Controls.Add(panelContenido);
            this.Controls.Add(panelPrincipal);

            // Centrar inicialmente
            CentrarContenido();
        }

        private void PanelPrincipal_Resize(object sender, EventArgs e)
        {
            CentrarContenido();
        }

        private void CentrarContenido()
        {
            if (panelContenido != null && panelPrincipal != null)
            {
                panelContenido.Location = new Point(
                    Math.Max(0, (panelPrincipal.Width - panelContenido.Width) / 2),
                    Math.Max(0, (panelPrincipal.Height - panelContenido.Height) / 2)
                );
            }
        }

        private void ConfigurarEventos()
        {
            btnConvertir.Click += BtnConvertir_Click;
            btnConvertir.MouseEnter += (s, e) => btnConvertir.BackColor = PURPURA_MONSTER;
            btnConvertir.MouseLeave += (s, e) => btnConvertir.BackColor = AZUL_CLARO;
        }

        private async void BtnConvertir_Click(object sender, EventArgs e)
        {
            try
            {
                double valor = double.Parse(txtValor.Text.Trim());
                int indice = cmbConversion.SelectedIndex;

                ConversionResponse resultado = null;

                switch (indice)
                {
                    case 0:
                        resultado = await clienteREST.MetrosAPiesAsync(valor);
                        break;
                    case 1:
                        resultado = await clienteREST.PiesAMetrosAsync(valor);
                        break;
                    case 2:
                        resultado = await clienteREST.MetrosAPulgadasAsync(valor);
                        break;
                    case 3:
                        resultado = await clienteREST.PulgadasAMetrosAsync(valor);
                        break;
                    case 4:
                        resultado = await clienteREST.KilometrosAMillasAsync(valor);
                        break;
                    case 5:
                        resultado = await clienteREST.MillasAKilometrosAsync(valor);
                        break;
                }

                if (resultado != null && resultado.Exito)
                {
                    string resultadoTexto = string.Format("{0:F2} {1} = {2:F2} {3}",
                        resultado.ValorOriginal, resultado.UnidadOrigen,
                        resultado.ValorConvertido, resultado.UnidadDestino);
                    MostrarResultado(resultadoTexto);
                }
                else
                {
                    MostrarError(resultado?.Mensaje ?? "Error desconocido");
                }
            }
            catch (FormatException)
            {
                MostrarError("Ingrese un valor numérico válido");
            }
            catch (Exception ex)
            {
                MostrarError("Error al enviar la solicitud: " + ex.Message);
            }
        }

        public void MostrarResultado(string resultado)
        {
            lblResultado.Text = resultado;
            lblResultado.ForeColor = AZUL_PRINCIPAL;
            lblResultado.BackColor = AZUL_CLARO;
            lblResultado.Font = new Font("Arial", 14F, FontStyle.Bold);
        }

        public void MostrarError(string error)
        {
            lblResultado.Text = "ERROR: " + error;
            lblResultado.ForeColor = ROJO_MONSTER;
            lblResultado.BackColor = Color.FromArgb(255, 240, 240);
            lblResultado.Font = new Font("Arial", 12F, FontStyle.Bold);
        }
    }
}

File: Proyectos/TI.1 PAO202650_SINBDD_NRC30742/TI.1.3.PAO202650_RESTFUL_JAVA_SINBDD_NRC30742/TI1.3 RESTFUL_JAVA_SINBDD_GR05/04.CLIESC/CONUNI_RESTFUL_JAVA_CLIESC_G05/src/main/java/ec/edu/monster/vista/PanelLongitud.java
package ec.edu.monster.vista;

import ec.edu.monster.servicio.ConversionApiClient;
import ec.edu.monster.modelo.ConversionResponse;
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

/**
 * Panel para conversiones de longitud - Diseño compacto y funcional
 * @author ACER NITRO V15
 */
public class PanelLongitud extends JPanel {
    
    // Colores oficiales de Monsters Inc.
    private static final Color AZUL_PRINCIPAL = new Color(33, 169, 218);      // #21a9da
    private static final Color ROJO_MONSTER = new Color(246, 105, 113);        // #f66971
    private static final Color AMARILLO_MONSTER = new Color(245, 216, 128);   // #f5d880
    private static final Color PURPURA_MONSTER = new Color(140, 107, 205);     // #8c6bcd
    private static final Color AZUL_CLARO = new Color(159, 220, 250);         // #9fdcfa
    private static final Color BLANCO = new Color(255, 255, 255);
    private static final Color GRIS_OSCURO = new Color(60, 60, 60);
    private static final Color GRIS_CLARO = new Color(240, 240, 240);
    
    // Componentes
    private JLabel lblTitulo;
    private JLabel lblValor;
    private JTextField txtValor;
    private JLabel lblConversion;
    private JComboBox<String> cmbConversion;
    private JButton btnConvertir;
    private JLabel lblResultado;
    
    // Cliente REST
    private ConversionApiClient clienteREST;
    
    public PanelLongitud() {
        clienteREST = new ConversionApiClient();
        inicializarComponentes();
        configurarLayout();
        configurarEventos();
    }
    
    private void inicializarComponentes() {
        // Título
        lblTitulo = new JLabel("CONVERSIONES DE LONGITUD");
        lblTitulo.setFont(new Font("Arial", Font.BOLD, 18));
        lblTitulo.setForeground(GRIS_OSCURO);
        lblTitulo.setHorizontalAlignment(SwingConstants.CENTER);
        
        // Valor
        lblValor = new JLabel("Valor a convertir:");
        lblValor.setFont(new Font("Arial", Font.BOLD, 14));
        lblValor.setForeground(GRIS_OSCURO);
        
        txtValor = new JTextField(15);
        txtValor.setFont(new Font("Arial", Font.PLAIN, 14));
        txtValor.setHorizontalAlignment(SwingConstants.CENTER);
        txtValor.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(GRIS_CLARO, 1),
            BorderFactory.createEmptyBorder(8, 12, 8, 12)
        ));
        
        // Conversión
        lblConversion = new JLabel("Tipo de conversión:");
        lblConversion.setFont(new Font("Arial", Font.BOLD, 14));
        lblConversion.setForeground(GRIS_OSCURO);
        
        String[] conversiones = {
            "Metros → Pies",
            "Pies → Metros", 
            "Metros → Pulgadas",
            "Pulgadas → Metros",
            "Kilómetros → Millas",
            "Millas → Kilómetros"
        };
        cmbConversion = new JComboBox<>(conversiones);
        cmbConversion.setFont(new Font("Arial", Font.PLAIN, 12));
        cmbConversion.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(GRIS_CLARO, 1),
            BorderFactory.createEmptyBorder(5, 8, 5, 8)
        ));
        
        // Botón convertir
        btnConvertir = new JButton("CONVERTIR");
        btnConvertir.setBackground(PURPURA_MONSTER);
        btnConvertir.setForeground(BLANCO);
        btnConvertir.setFont(new Font("Arial", Font.BOLD, 16));
        btnConvertir.setPreferredSize(new Dimension(200, 45));
        btnConvertir.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(PURPURA_MONSTER, 1),
            BorderFactory.createEmptyBorder(12, 25, 12, 25)
        ));
        btnConvertir.setFocusPainted(false);
        btnConvertir.setCursor(new Cursor(Cursor.HAND_CURSOR));
        
        // Resultado
        lblResultado = new JLabel("Resultado aparecerá aquí");
        lblResultado.setFont(new Font("Arial", Font.BOLD, 16));
        lblResultado.setForeground(GRIS_OSCURO);
        lblResultado.setHorizontalAlignment(SwingConstants.CENTER);
        lblResultado.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(GRIS_CLARO, 1),
            BorderFactory.createEmptyBorder(15, 20, 15, 20)
        ));
        lblResultado.setBackground(GRIS_CLARO);
        lblResultado.setOpaque(true);
    }
    
    private void configurarLayout() {
        setBackground(BLANCO);
        setLayout(new GridBagLayout());
        GridBagConstraints gbc = new GridBagConstraints();
        
        // Título
        gbc.gridx = 0; gbc.gridy = 0; gbc.gridwidth = 2;
        gbc.insets = new Insets(20, 0, 30, 0);
        add(lblTitulo, gbc);
        
        // Valor
        gbc.gridx = 0; gbc.gridy = 1; gbc.gridwidth = 1;
        gbc.anchor = GridBagConstraints.EAST;
        gbc.insets = new Insets(10, 20, 10, 10);
        add(lblValor, gbc);
        
        gbc.gridx = 1; gbc.gridy = 1;
        gbc.anchor = GridBagConstraints.WEST;
        gbc.insets = new Insets(10, 10, 10, 20);
        add(txtValor, gbc);
        
        // Conversión
        gbc.gridx = 0; gbc.gridy = 2;
        gbc.anchor = GridBagConstraints.EAST;
        gbc.insets = new Insets(10, 20, 10, 10);
        add(lblConversion, gbc);
        
        gbc.gridx = 1; gbc.gridy = 2;
        gbc.anchor = GridBagConstraints.WEST;
        gbc.insets = new Insets(10, 10, 10, 20);
        add(cmbConversion, gbc);
        
        // Botón
        gbc.gridx = 0; gbc.gridy = 3; gbc.gridwidth = 2;
        gbc.anchor = GridBagConstraints.CENTER;
        gbc.insets = new Insets(20, 0, 20, 0);
        add(btnConvertir, gbc);
        
        // Resultado
        gbc.gridx = 0; gbc.gridy = 4; gbc.gridwidth = 2;
        gbc.fill = GridBagConstraints.HORIZONTAL;
        gbc.insets = new Insets(20, 20, 20, 20);
        add(lblResultado, gbc);
    }
    
    private void configurarEventos() {
        btnConvertir.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                convertirLongitud();
            }
        });
        
        // Efecto hover en botón
        btnConvertir.addMouseListener(new java.awt.event.MouseAdapter() {
            public void mouseEntered(java.awt.event.MouseEvent evt) {
                btnConvertir.setBackground(AZUL_PRINCIPAL);
            }
            public void mouseExited(java.awt.event.MouseEvent evt) {
                btnConvertir.setBackground(PURPURA_MONSTER);
            }
        });
    }
    
    private void convertirLongitud() {
        try {
            double valor = Double.parseDouble(txtValor.getText().trim());
            int indice = cmbConversion.getSelectedIndex();
            
            String operacion = "";
            
            switch (indice) {
                case 0: 
                    operacion = "metrosAPies"; 
                    break;
                case 1: 
                    operacion = "piesAMetros"; 
                    break;
                case 2: 
                    operacion = "metrosAPulgadas"; 
                    break;
                case 3: 
                    operacion = "pulgadasAMetros"; 
                    break;
                case 4: 
                    operacion = "kilometrosAMillas"; 
                    break;
                case 5: 
                    operacion = "millasAKilometros"; 
                    break;
            }
            
            // Realizar conversión usando el servicio REST
            ConversionResponse conversion = clienteREST.convertirLongitud(operacion, valor);
            
            if (conversion.isExito()) {
                String resultadoTexto = String.format("%.2f %s = %.2f %s", 
                    conversion.getValorOriginal(), conversion.getUnidadOrigen(), 
                    conversion.getValorConvertido(), conversion.getUnidadDestino());
                mostrarResultado(resultadoTexto);
            } else {
                mostrarError(conversion.getMensaje());
            }
            
        } catch (NumberFormatException ex) {
            mostrarError("ERROR: Ingrese un valor numérico válido");
        }
    }
    
    
    public void mostrarResultado(String resultado) {
        // Mostrar resultado localmente (abajo del botón)
        lblResultado.setText(resultado);
        lblResultado.setForeground(PURPURA_MONSTER);
        lblResultado.setBackground(AZUL_CLARO);
        lblResultado.setOpaque(true);
        lblResultado.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(PURPURA_MONSTER, 1),
            BorderFactory.createEmptyBorder(10, 15, 10, 15)
        ));
        lblResultado.setFont(new Font("Arial", Font.BOLD, 14));
    }
    
    public void mostrarError(String error) {
        // Mostrar error localmente (abajo del botón)
        lblResultado.setText("ERROR: " + error);
        lblResultado.setForeground(ROJO_MONSTER);
        lblResultado.setBackground(new Color(255, 240, 240));
        lblResultado.setOpaque(true);
        lblResultado.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(Color.RED, 1),
            BorderFactory.createEmptyBorder(10, 15, 10, 15)
        ));
        lblResultado.setFont(new Font("Arial", Font.BOLD, 14));
    }
}